fabtests/unit/mr_cache_evict.c

```diff
@@ -0,0 +1,695 @@
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <unistd.h>
+#include <stddef.h>
+#include <sys/mman.h>
+#include <getopt.h>
+#include <time.h>
+
+#ifdef __linux__
+#include <malloc.h>
+#endif
+
+#include "unit_common.h"
+#include "shared.h"
+
+/* Supported memory region types. */
+enum mem_type {
+	MMAP,
+	BRK,
+	SBRK,
+};
+
+struct memory_region {
+	enum mem_type type;
+	size_t size;
+	char *user;
+};
+
+/* Given a time value in seconds, determine the expected cached time value. The
+ * assumption is that the cache value should at least have a
+ * CACHE_IMPROVEMENT_PERCENT improvement over the original value.
+ */
+#define CACHE_IMPROVEMENT_PERCENT 0.8
+#define CACHE_TIME_MAX_VALUE(time) ((time) * (1 - CACHE_IMPROVEMENT_PERCENT))
+
+#define PGMAP_PRSNT (1UL << 63)
+#define PFN_MASK ((1UL << 55) - 1)
+#define FOFFSET(va) ((((uint64_t)va) >> 12) << 3)
+
+static void *reuse_addr = NULL;
+static char err_buf[512];
+static size_t mr_buf_size = 4097;
+
+/* Function used to get physical address from a virtual address. Must be root
+ * to read pagemap.
+ */
+static int virt_to_phys(const void *va_addr, uint64_t *phy_addr)
+{
+	FILE *fd;
+	int ret;
+	uint64_t entry;
+	size_t read_size;
+
+	fd = fopen("/proc/self/pagemap", "rb");
+	if (!fd) {
+		ret = -errno;
+		FT_UNIT_STRERR(err_buf, "fopen of /proc/self/pagemap failed",
+			       ret);
+		return ret;
+	}
+
+	ret = fseek(fd, FOFFSET(va_addr), 0);
+	if (ret == -1) {
+		FT_UNIT_STRERR(err_buf, "fseek failed", ret);
+		goto out;
+	}
+
+	read_size = fread(&entry, sizeof(entry), 1, fd);
+	if (read_size != 1) {
+		if (feof(fd)) {
+			ret = -ENOSPC;
+			FT_UNIT_STRERR(err_buf, "Premature EOF", ret);
+			goto out;
+		}
+
+		/* Instead of explicitly checking ferror(), just assume it is
+		 * set.
+		 */
+		ret = -EIO;
+		FT_UNIT_STRERR(err_buf, "fread error", ret);
+		goto out;
+	}
+
+	if (entry & PGMAP_PRSNT) {
+		ret = 0;
+		*phy_addr = (entry & PFN_MASK) << 12;
+	} else {
+		ret = -EFAULT;
+		FT_UNIT_STRERR(err_buf, "Failed to find physical address", ret);
+	}
+
+out:
+	fclose(fd);
+
+	return ret;
+}
+
+/* Sbrk/brk allocations are only intended to support a single outstanding
+ * allocation at a time. Extra handling of the program break is needed to make
+ * sbrk/brk allocations more flexible including making allocations thread safe.
+ */
+static void sbrk_free(struct memory_region *mr)
+{
+	void *cur_brk = (void *)((uint64_t)mr + mr->size);
```

Comment:
nit: throughout the code, please add a space after all casts.  e.g. (uint64_t) mr, (void *) -1, etc.  This matches the coding style that is usually used throughout the codebase.

Reply:
Will do.